Properly handle unset partition-replication in restatectl

muhamadazmy opened the pull request:
Properly handle unset partition-replication in restatectl

Summary:
Ensure correct behavior when the partition-replication flag is omitted in both provision and config set.
Previously, an unset flag defaulted to everywhere, which was incorrect.

- In provision, if the flag is omitted, the cluster defaults to the configured default-partition-replication.
 The user can override this by specifying a valid replication property.

 - In config set, omitting the flag leaves the current cluster configuration unchanged.

tillrohrmann commented on crates/core/protobuf/node_ctl_svc.proto:
```diff
@@ -34,9 +34,9 @@ message ProvisionClusterRequest {
-  // if unset partition replication will default to `everywhere`. Otherwise
-  // it's limited to the provided replication property
-  optional restate.cluster.ReplicationProperty partition_replication = 3;
+  // if unset partition replication will default to the configured
+  // `default-partition-replication`
+  optional restate.cluster.PartitionReplication partition_replication = 3;
```
Do we need to keep the old field for ensuring backward compatibility with older restatectl versions? Would a new `restatectl` version work with an older server?

tillrohrmann commented on tools/restatectl/src/commands/config.rs:
```diff
@@ -43,8 +44,14 @@ pub fn cluster_config_string(config: &ClusterConfiguration) -> anyhow::Result<St
+                replication_property.replication_property.as_str()
+            }
+        })
+        .unwrap_or("unknown"); // should never happen!
```
Should we make it an `expect` if it should never happen?

tillrohrmann commented on crates/types/protobuf/restate/cluster.proto:
```diff
@@ -92,8 +92,15 @@ message BifrostProvider {
 }
 
+message PartitionReplication {
+  oneof replication {
+    ReplicationProperty limited = 1;
+    google.protobuf.Empty everywhere = 2;
+  }
```
It's not super critical because we don't use `ClusterConfiguration` on the hot path, but I wanted to share that `oneofs` usually lead to deeper nested structures which can impact serde performance negatively.

muhamadazmy replied:
I was trying to avoid adding an extra fields directly on the ClusterConfiguration since it will feel out of place and allows both fields (partition_replication and everywhere) to be set at the same time.

But it seems i used oneof in the wrong way here for compatibility. I should have instead replaced the partition_replication in-place instead of creating a new message that encapsulate the oneof! sorry about that.

So maybe I will just add an enum instead

AhmedSoliman replied:
YMMV, but oneof is not different from encoding multiple attributes as messages. I wouldn't be worried about it as long as the nesting is not incredibly deep.

tillrohrmann reviewed (commented):
Thanks a lot for fixing setting the partition replication @muhamadazmy. The changes look good to me. I had one question regarding backwards/forward compatibility. Maybe we can keep it by introducing a new partition replication field and keeping the old one.

muhamadazmy commented:
Thanks @tillrohrmann for your review.

I updated the PR to maintain both backward and forward compatibility. Also dropped the use of `oneof` and will avoid it in the future. This fix also allows us to extend the kind of partition replication in the future as needed.

Please take another look and let me know if there is any suggestions

muhamadazmy commented:
@tillrohrmann, I have modified the PR here as we agreed to phase out "everywhere" as a valid partition replication policy. While the policy itself is still supported by the system for backward compatibility the user has no way to choose it during `provision` or via `config set`

tillrohrmann commented on tools/restatectl/src/commands/provision.rs:
```diff
@@ -30,9 +30,8 @@ pub struct ProvisionOpts {
-    /// Optional partition placement strategy. By default replicates
-    /// partitions on all nodes. Accepts replication property
-    /// string as a value
+    /// Optional partition placement strategy. If unset, uses
```
"Partition replication. If unset, ..."

tillrohrmann reviewed (approved):
Thanks a lot for fixing setting the partition replication @muhamadazmy. LGTM. +1 for merging :-)
